### PR TITLE
Fill context with the information on which route triggered the function.

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Based on https://github.com/vaadin/vaadin-router/pull/137

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/138)

### CODE REVIEW

**Comment by SomeoneToIgnore** [comment]
*[test/router/dynamic.redirect.spec.html, line 289 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/138#-LEBWunm79eppjqQJ-4W:-LEBWunm79eppjqQJ-4X:ba4y3j5) ([raw file](https://github.com/vaadin/vaadin-router/blob/f7456b821e77e5dab7215d1382df7d33e4671e61/test/router/dynamic.redirect.spec.html#L289)):*
> ```HTML
>       });
> 
>       describe('redirect.from and context.from values for dynamic redirects', () => {
> ```

2 more methods (for `onBeforeEnter` and `onBeforeLeave`) will be added in future.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/138)*

**Comment by SomeoneToIgnore** [comment]
Before this PR, 
> ```
> inactivatedRoute
> ```
was used to store the information about the route that had triggered the `inactivate` function (this is not the same route as the one currently in resolution, that's why it's needed).

`onBeforeLeave` has more or less the same requirement but it's not `inactivatedRoute` anymore.
That's why I've decided to name the variable more abstract.

In addition, `inactivatedRoute` was filled incorrectly in the redirect case, the issue is fixed in the PR.

The PR is based on https://github.com/vaadin/vaadin-router/pull/137, that one should be reviewed first.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/138#-:-LEBX-HcErd5-G4xr0e3:b-et2khg)*

**Comment by SomeoneToIgnore** [comment]
> The PR is based on #137, that one should be reviewed first.

Not true anymore, ready to be reviewed.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/138#-:-LEGoDoQEI84dxx9EOtV:b-dt924e)*

**Comment by web-padawan** [comment]
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[src/router.js, line 41 at r3](https://reviewable.io/reviews/vaadin/vaadin-router/138#-LEJ2Jk8DPVtJsVPaJ9x:-LEJ2Jk8DPVtJsVPaJ9y:b-4aeiyi) ([raw file](https://github.com/vaadin/vaadin-router/blob/397a513fd1a05412a7abe6c8f4900f1ddb3680e4/src/router.js#L41)):*
> ```JavaScript
> function runCallbackIfPossible(callback, context, invocationPath) {
>   if (typeof callback === 'function') {
>     return callback(Object.assign({}, context, {invocationPath: invocationPath}));
> ```

nit: you can use ES6 object shorthands:

```
Object.assign({}, context, {invocationPath})
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/138)*

**Comment by SomeoneToIgnore** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/router.js, line 41 at r3](https://reviewable.io/reviews/vaadin/vaadin-router/138#-LEJ2Jk8DPVtJsVPaJ9x:-LEJ4LcU-tgR5pcxKIO5:b-1uq9is) ([raw file](https://github.com/vaadin/vaadin-router/blob/397a513fd1a05412a7abe6c8f4900f1ddb3680e4/src/router.js#L41)):*
<details><summary><i>Previously, web-padawan (Sergey Kulikov) wrote…</i></summary><blockquote>

nit: you can use ES6 object shorthands:

```
Object.assign({}, context, {invocationPath})
```
</blockquote></details>

Awesome, fixed.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/138)*

**Comment by vlukashov** [comment]
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[src/router.js, line 286 at r4](https://reviewable.io/reviews/vaadin/vaadin-router/138#-LEJJowHEMso8nTELAgm:-LEJJowHEMso8nTELAgn:bamjo5g) ([raw file](https://github.com/vaadin/vaadin-router/blob/42ca04926ae6b614e549dba7b9f43c03589ef73f/src/router.js#L286)):*
> ```JavaScript
>    * 'context' parameter can be used for asynchronously getting the next route contents in the resolution chain via 'context.next()'
>    * and for getting route parameters via 'context.params'.
>    * Also, 'context' contains 'invocationPath' property that holds the info on the route that caused the function to be triggered.
> ```

For the `action` callback the `invocationPath` property it is not important, and no other purpose is documented at all. Should we either keep the property private, or document other usages?

It's clear that _we_ need this property to implement `redirect.from` correctly, but I do not see yet why can it be useful for end users.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/138)*

**Comment by SomeoneToIgnore** [comment]
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/router.js, line 286 at r4](https://reviewable.io/reviews/vaadin/vaadin-router/138#-LEJJowHEMso8nTELAgm:-LEJQs_WDekFiD4qAfMA:b-izo6cs) ([raw file](https://github.com/vaadin/vaadin-router/blob/42ca04926ae6b614e549dba7b9f43c03589ef73f/src/router.js#L286)):*
<details><summary><i>Previously, vlukashov (Viktor Lukashov) wrote…</i></summary><blockquote>

For the `action` callback the `invocationPath` property it is not important, and no other purpose is documented at all. Should we either keep the property private, or document other usages?

It's clear that _we_ need this property to implement `redirect.from` correctly, but I do not see yet why can it be useful for end users.
</blockquote></details>

It is useful in two cases: `onBeforeLeave` and `inactivate` – in both cases the `context.route` !== `context. invocationPath`

Have updated the doc.

Also renamed it to `invocationRoute`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/138)*

**Comment by SomeoneToIgnore** [comment]
*[src/router.js, line 286 at r4](https://reviewable.io/reviews/vaadin/vaadin-router/138#-LEJJowHEMso8nTELAgm:-LEJoz_-5s_X5s8k25PQ:b-rrf3h7) ([raw file](https://github.com/vaadin/vaadin-router/blob/42ca04926ae6b614e549dba7b9f43c03589ef73f/src/router.js#L286)):*
<details><summary><i>Previously, SomeoneToIgnore (Kirill Bulatov) wrote…</i></summary><blockquote>

It is useful in two cases: `onBeforeLeave` and `inactivate` – in both cases the `context.route` !== `context. invocationPath`

Have updated the doc.

Also renamed it to `invocationRoute`.
</blockquote></details>

Reworked as discussed.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/138)*

**Comment by vlukashov** [comment]
Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[src/router.js, line 244 at r6](https://reviewable.io/reviews/vaadin/vaadin-router/138#-LEJqjrHBGF6-SkgJ_kl:-LEJqjrHBGF6-SkgJ_km:bwuxgds) ([raw file](https://github.com/vaadin/vaadin-router/blob/19d6760206d0fea55af24bd86c52092e8ca4ac90/src/router.js#L244)):*
> ```JavaScript
>    *
>    *
>    * `context` objet that is passed to `route` functions holds the following parameters:
> ```

Now it looks like the same info is repeated in two places: here in the API docs, and in the "Route Actions" live demos. Would it make sense to sync these two?

Now the API docs seem to be missing the info on `context.pathname`, `context.redirect()`, `context.component()`, and `context.cancel()`.

---

*[src/router.js, line 249 at r6](https://reviewable.io/reviews/vaadin/vaadin-router/138#-LEJsXMM2Hkm9_0Kd4g_:-LEJsXMM2Hkm9_0Kd4ga:b-14iam8) ([raw file](https://github.com/vaadin/vaadin-router/blob/19d6760206d0fea55af24bd86c52092e8ca4ac90/src/router.js#L249)):*
> ```JavaScript
>    *  * `route` that holds the route that is currently being rendered.
>    *
>    * If the function defined not as an arrow one but with the `function` keyword, its `this` property points to the route that had
> ```

I'm not sure if `the function that is an arrow one` is clear enough. How about rewording this to something along the lines of

_The original `route` object that defined an `action` or `inactivate` callback is available inside the callback through the `this` reference.  The current route object being currently resolved is available there as `context.route`. While for `action` callbacks these are always the same, for `inactivate` callbacks these objects would be different: the current route that is being rendered causes another route to inactivate. If you need to access the original route object, make sure you define the `inactivate` callback as a non-arrow function because arrow functions do not have their own `this` reference._

That also removes all references to `invocationRoute`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/138)*

**Comment by SomeoneToIgnore** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/router.js, line 244 at r6](https://reviewable.io/reviews/vaadin/vaadin-router/138#-LEJqjrHBGF6-SkgJ_kl:-LEJued93Bfzd9dxC53g:b-1eh3wi) ([raw file](https://github.com/vaadin/vaadin-router/blob/19d6760206d0fea55af24bd86c52092e8ca4ac90/src/router.js#L244)):*
<details><summary><i>Previously, vlukashov (Viktor Lukashov) wrote…</i></summary><blockquote>

Now it looks like the same info is repeated in two places: here in the API docs, and in the "Route Actions" live demos. Would it make sense to sync these two?

Now the API docs seem to be missing the info on `context.pathname`, `context.redirect()`, `context.component()`, and `context.cancel()`.
</blockquote></details>

Sure, done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/138)*

**Comment by SomeoneToIgnore** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/router.js, line 249 at r6](https://reviewable.io/reviews/vaadin/vaadin-router/138#-LEJsXMM2Hkm9_0Kd4g_:-LEJy06UDvHy0IvUC6YE:b-noe1ja) ([raw file](https://github.com/vaadin/vaadin-router/blob/19d6760206d0fea55af24bd86c52092e8ca4ac90/src/router.js#L249)):*
<details><summary><i>Previously, vlukashov (Viktor Lukashov) wrote…</i></summary><blockquote>

I'm not sure if `the function that is an arrow one` is clear enough. How about rewording this to something along the lines of

_The original `route` object that defined an `action` or `inactivate` callback is available inside the callback through the `this` reference.  The current route object being currently resolved is available there as `context.route`. While for `action` callbacks these are always the same, for `inactivate` callbacks these objects would be different: the current route that is being rendered causes another route to inactivate. If you need to access the original route object, make sure you define the `inactivate` callback as a non-arrow function because arrow functions do not have their own `this` reference._

That also removes all references to `invocationRoute`.
</blockquote></details>

`invocationRoute` references were left by the mistake.

Thanks, updated the docs with the proposed sentence.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/138)*

**Comment by vlukashov** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/router.js, line 22 at r7](https://reviewable.io/reviews/vaadin/vaadin-router/138#-LEK0SISBTZrdTwLj2Fs:-LEK0SIT8AIhr26lD3H0:bd98n5g) ([raw file](https://github.com/vaadin/vaadin-router/blob/31f2d9b042d8f680aa4c1862934edbec94736596/src/router.js#L22)):*
> ```JavaScript
>   let currentPath = context.__invocationRoute;
>   let invocationPathName = '';
>   while (currentPath) {
> ```

This way of calculating the invocation pathname would break on the edge cases with `''` and `'/'` routes in the chain:

E.g. this test case would fail:
```
it('action with redirect has its context.from and redirect.from populated correctly (2)', async() => {
  let from;
  router.setRoutes([
    {path: '/', children: [
      {path: '', children: [
        {path: '/c', action: context => context.redirect('/d')}
      ]},
    ]},
    {path: '/d', component: 'x-home-view', action: context => {from = context.from}}
  ]);

  await router.render('/c');

  expect(outlet.children[0].tagName).to.match(/x-home-view/i);
  expect(from).to.be.equal('/c');
});
```

---

*[src/router.js, line 113 at r7](https://reviewable.io/reviews/vaadin/vaadin-router/138#-LEK0lyL6vASLw7PtaX6:-LEK0lyL6vASLw7PtaX7:b-sjx8zp) ([raw file](https://github.com/vaadin/vaadin-router/blob/31f2d9b042d8f680aa4c1862934edbec94736596/src/router.js#L113)):*
> ```JavaScript
> 
>     context.redirect = function(path) {
>       return redirect(this, path);
> ```

This test case would fail:
```
it('context.redirect() should work when invoked without the `this` context', async() => {
  router.setRoutes([
    {path: '/', component: 'x-home-view'},
    {path: '/a', action: ({redirect}) => redirect('/d')}
  ]);

  await router.render('/a');

  expect(outlet.children[0].tagName).to.match(/x-home-view/i);
});
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/138)*
